napari/viewer.py

```diff
@@ -140,6 +140,8 @@ def show(self, *, block=False):
 
     def close(self):
         """Close the viewer window."""
+        # Shutdown the slicer first to avoid processing any more tasks.
+        self._layer_slicer.shutdown()
         # Remove all the layers from the viewer
         self.layers.clear()
```

Comment:
This will cause exceptions to be raised from `LayerSlicer` when there is more than 1 layer. I think there are a few options here, but probably disconnecting the layer slicer from events that trigger slicing seems like the best option here.

Reply:
The exception is being thrown in the done callback when calling `Future.result`, but the shutdown runtimeerror is actually coming from dask:

<details>

```
exception calling callback for <Future at 0x7f8795114370 state=finished raised RuntimeError>
Traceback (most recent call last):
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/concurrent/futures/_base.py", line 329, in _invoke_callbacks
    callback(self)
  File "/Users/asweet/projects/napari/napari/components/_layer_slicer.py", line 244, in _on_slice_done
    result = task.result()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/concurrent/futures/_base.py", line 438, in result
    return self.__get_result()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/asweet/projects/napari/napari/components/_layer_slicer.py", line 230, in _slice_layers
    return {layer: request() for layer, request in requests.items()}
  File "/Users/asweet/projects/napari/napari/components/_layer_slicer.py", line 230, in <dictcomp>
    return {layer: request() for layer, request in requests.items()}
  File "/Users/asweet/projects/napari/napari/layers/image/_slice.py", line 88, in __call__
    self._call_multi_scale()
  File "/Users/asweet/projects/napari/napari/layers/image/_slice.py", line 156, in _call_multi_scale
    image = np.asarray(image)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/array/core.py", line 1689, in __array__
    x = self.compute()
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/base.py", line 315, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/base.py", line 603, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/threaded.py", line 89, in get
    results = get_async(
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/local.py", line 499, in get_async
    fire_tasks(chunksize)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/site-packages/dask/local.py", line 494, in fire_tasks
    fut = submit(batch_execute_tasks, each_args)
  File "/Users/asweet/software/miniconda3/envs/napari-dev/lib/python3.9/concurrent/futures/thread.py", line 161, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
RuntimeError: cannot schedule new futures after shutdown
```

</details>

I'm not sure what causes the dask executor to shutdown. Also not sure if we can control it. If we can at least understand if it's shutdown in a thread safe way, we may be able to mitigate this cleanly.

A cruder fix could  be to prevent reraising any exceptions after shutdown (e.g. by storing that state and skipping the done callback when shutdown).

Reply:
We are going to be touching the Dask code very soon, specifically for executors. 

That said, I poked a bit and it looks like the issue may be due to Dask and napari both using the same executor pool, and napari shutting it down before Dask is ready. https://github.com/dask/distributed/issues/6087#issuecomment-1092615183. For tiled loading we want to manage our own dask executor.

Reply:
Thanks for the link. As from the architecture working group, I understood that the CPython interpreter will shutdown all executors before the interpreter exits. By I don't see how shutting down our executor could shutdown dask's - the very idea still terrifies me! Do you have a more specific link for that?

My guess is that this is more related to the order in which CPython shuts down the executors at exit. I.e. if the dask executor is shutdown before ours, there's a chance we'll call `asarray` on our executor's thread. This chance is increased since we wait for futures to finish on shutdown - we could pass `wait=False` to decrease this chance, but I think it could still happen.